numpy/core/src/multiarray/ctors.c

```diff
@@ -1876,7 +1882,17 @@ PyArray_GetArrayParamsFromObject(PyObject *op,
             return 0;
         }
         /* If object arrays are forced */
-        if (is_object) {
+        if (is_object != DISCOVERED_OK) {
+            if (is_object == DISCOVERED_RAGGED && requested_dtype == NULL) {
+                /* NumPy 1.18, 2019-11-01 */
+                if (DEPRECATE("Creating an ndarray with automatic object "
+                    "dtype is deprecated, use dtype=object if you intended "
+                    "it, otherwise specify an exact dtype") < 0)
```

Comment:
could style nit, that the indentation should be deeper, but please ignore if you want.

Reply:
Will try to sneak it into another PR